### ISSUE TITLE
The matching strategy is flawed

### ISSUE THREAD

**Issue by aradalvand** [meilisearch/meilisearch#4093]
### Describe the bug
Hi there.

I feel like there must be something I'm missing because this doesn't make any sense at all. If so then please let me know. Thanks.

If you have a document with a searchable field containing the string "Apple MacBook", and you run a search against the index with the query `appl macbook` (a pretty realistic typo, for example), Meilisearch fails to return that document as a hit.

**To Reproduce**

1. First create a document in a new index like this (so with default settings):
```sh
curl -X POST http://localhost:7700/indexes/products/documents \
  -H 'Authorization: Bearer <api-key>' \
  -H 'Content-Type: application/json' \
  --data '{
    "id": 1,
    "title": "Apple MacBook"
  }'
```

2. Then run this search:
```sh
curl -X POST http://localhost:7700/indexes/products/search \
  -H 'Authorization: Bearer <api-key>' \
  -H 'Content-Type: application/json' \
  --data '{
    "q": "appl macbook"
  }'
```

There will be 0 hits:

![image](https://github.com/meilisearch/meilisearch/assets/26527405/9a6cb4b3-a072-432e-a35c-9579ee11b80a)

Needless to say running a query without the typo does return the document:

![image](https://github.com/meilisearch/meilisearch/assets/26527405/4286cd77-fc05-4b0b-873c-6459b9af03b8)

**Expected behavior**
Where is the typo-tolerance?! Am I doing anything wrong? This is one of the most basic tests for typo-tolerance one could possibly come up with, so I don't understand Meilisearch's behavior here. Is this a known bug?

**Meilisearch version:**
1.3.5

**Comment by aradalvand** [comment]
**Update:** After some more investigation, it seems like the problem is deeper than just typo-tolerance.

If you have a product with a `title` of `Apple MacBook Pro 2019`, and you run a search against this index with the query `appl macbook pro 2019`, it returns NO results; which is obviously unacceptable; literally 99% of the product title is included in the query VERBATIM, and Meilisearch still doesn't return it!

Correct me if I'm wrong but it seems to me that this has something to do with the [default matching strategy](https://www.meilisearch.com/docs/reference/api/search#matching-strategy) that Meilisearch uses, where if there are not enough documents that match the entire query, Meilisearch will remove words from the **end** of the query one-by-one to find a match and compare the result against the documents again.

But unfortunately, the only other matching strategy that you can pick is `all` which is even more strict and needs a given document to match the entire query in order to return it as a hit.

There needs to be a third option — it can be called `firstAndLast` or `bidirectional` or something like that — that is similar to `last` but does the removing of words in both directions, namely both from the end and also from the start of the query.

**Comment by CaroFG** [comment]
Hello!  It looks like the focus here isn't on matching strategy, but rather on typo tolerance. Just a little heads-up: by default, Meilisearch allows for one typo in query terms that are five or more characters long. Feel free to tailor those defaults to meet your unique needs! The [documentation on typo tolerance](https://www.meilisearch.com/docs/learn/configuration/typo_tolerance#configuring-typo-tolerance) is here to help guide you. 📚

**Comment by aradalvand** [comment]
@CaroFG It is **not**, sorry, but did you even bother to read my explanation?! I literally talked about this.

If it was about typo-tolerance per se, then a query like `apple macbk pr 201` should've not matched the document, but it currently does; yet another query with a lot fewer typos like `appl macbook pro 2019` does NOT. This is not just about typo-tolerance.

What justification could there possibly be for a query that matches a document almost completely verbatim to not result in any hits whatsoever?! 

And again, the `Support` tag is not appropriate for this issue. I'm describing a blatant problem.

**Comment by CaroFG** [comment]
Thank you so much for sharing your example with us! It seems like the matching strategy is working as intended. However, I totally understand that it doesn't quite fit your use case. It sounds like you’re suggesting a new feature to better meet your needs. I’m excited to share your idea with our product team by moving this to our product repository as a feature proposal. Your insights are super valuable and we appreciate you taking the time to help us improve!

**Comment by aradalvand** [comment]
This is not really about "a new feature to better meet by needs", I'm more inclined to say the current behavior is problematic enough to merit being called a bug. Again, let me reiterate the crux of the issue: Meilisearch fails to return a document with a field with the value `Apple MacBook Pro 2019` when the query `Appl MacBook Pro 2019` is searched. This simply makes no sense. Meilisearch alreadyt promises typo-tolerance, but fails to fulfill that promise adequately in scenarios like this one.

Typesense, for example, is much saner in this regard — see [the docs](https://typesense.org/docs/0.25.1/api/search.html#typo-tolerance-parameters:~:text=tolerance.%0A%0ADefault%3A%201-,drop_tokens_threshold,-no) — I quote (emphasis mine):

> If drop_tokens_threshold is set to a number N and a search query contains multiple words (eg: wordA wordB), if at least N results with both wordA and wordB in the same document are not found, then Typesense will drop wordB and search for documents with just wordA. Typesense will keep dropping keywords like this **left to right and/or right to left**, until at least N documents are **found.**

So, just like what I described in my second comment, Typesense drops words/tokens in both directions, which is how it should be done. Meilisearch, on the other hand, is not doing enough.

Also, [this gist](https://gist.github.com/kishorenc/126c5529c9c65df7133ec087928eaada) showcases the same `Apple MacBook Pro 2019` example that I alluded to multiple times working as expected in Meilisearch.

**Comment by aradalvand** [comment]
@Kerollmops @irevoire @qdequele Sorry for the ping but could any of you guys please confirm this? Thanks.

**Comment by aradalvand** [comment]
This was pointed out [here](https://github.com/meilisearch/product/discussions/97) almost two years ago and it was hand-waved away like it's nothing... Unbelievable...

**Comment by azngeek** [comment]
I also do not understand the expected behaviour. The blog https://blog.meilisearch.com/how-full-text-search-engines-work/ outlines, that Meilisearch should be capable of applying fuzzy-like search on a search string with multiple search terms. 

I would expect this to also work.

Btw: Our current workaround is to split all the search term combination in multiple search requests using multi-search and then do our own calculation. This is just a really bad workaround.

**Comment by aradalvand** [comment]
It's been a week since I created this issue and I have gotten no proper response so far. This issue #4103 was created a few hours ago and got a response by @qdequele is ONE HOUR.

Why?!

**Comment by aradalvand** [comment]
@curquiza @gmourier I"m still waiting for a response...

**Comment by aradalvand** [comment]
@curquiza @gmourier @ManyTheFish I'm still waiting for a response...

**Comment by aradalvand** [comment]
(Closed the issue by accident)

@curquiza @gmourier @ManyTheFish I'm still waiting for a response...

**Comment by khalibloo** [comment]
I ran into a similar scenario where "whit hoodie" just would not match "white hoodie". I think it's a bug given that any other fuzzy matching system I know would have matched that.

That said, while I understand the frustration to some extent, it really doesn't help to be this pushy about a bug found in software. Paid or open source. Bugs are to be expected in any software. Everyone's just doing their best to help where they can. No one has the obligation to respond within 1 hour. This will just make it less likely that anyone would want to volunteer to walk into the line of fire.

Anyway, for the maintainers, it seems the fuzzy matching doesn't work as well when it's the first word that has the typo. It doesn't seem to have a problem matching when the typo is in the 2nd word.

**Comment by aradalvand** [comment]
As I explicitly pointed out on the Discord server, this project is not what it advertises itself to be. It markets itself as an alternative to actually-capable pieces of software such as ElasticSearch, Algolia, and Typesense, whereas the reality of the situation is radically different.

It is revealed all too quickly, once one examines this project with sufficient depth, that it is, in fact, a terrible piece of software, with astoundingly incompetent maintainers who have demonstrated themselves to be not only unresponsive and unreceptive to criticism, but also extraordinarily irresponsible. I was told on Discord, upon mildly complaining about not having received any proper reaction from the maintainers after more than a week of reporting what is the most obvious imaginable flaw in a fuzzy search engine  — let's not forget that Meilisearch is, after all, a company with full-time engineers working on the software, it's not like people are spending their free time on this thing, which would've made it somewhat unreasonable to expect responsiveness — that the more "aggressively" I pursue this, the less likely I am to get an answer, as if they're talking to a child that needs to "behave", and as if they would be doing _me_ a _favor_ by fixing a blatant bug that I have taken the time to thoroughly describe. This attitude is showcases lack of professionalism and seriousness, it is very, very telling.

It's patently clear at this point that the maintainers do not really care about the quality of their product — exemplified by their dismissive attitude upon being informed about a blatant defect in their software, and their resorting to childish  — and frankly that's nothing new for a VC-backed garbageware. If they had spent 1/100th of the effort they've spent on marketing and making their website's UI pretty, on the core functionality of their product, it wouldn't have been the embarrassment that it is.

I have personally migrated to Typesense, and I'd recommend everyone who encounters this issue to do the same as well. The sheer disparity between the professionalism and seriousness shown by the folks at Typesense and those that work on this project (or so they pretend to), is well and truly startling.

**Comment by irevoire** [comment]
Hey,
@curquiza I think we can officially ban @aradalvand from GitHub and Discord, who clearly didn't respect our code of conduct. And when I told him that his comments were becoming disrespectful, he directly [insulted us on discord](https://discord.com/channels/1006923006964154428/1006923007559729154/1160268900726997053).

-------

For the other one, as @aradalvand noted, this is actually the expected behavior; that’s how meilisearch is supposed to behave with the `matching_strategy: last`. We understand that it may not be perfect, and we already prioritized implementing other strategies for the v1.3.
But since almost no one asked for it, it wasn’t high on our roadmap, and we did not have the time to fix it.
However, we've been tracking how many people are asking for it to help us prioritize it.

@khalibloo, if you have the time, it would be very nice to post a comment in the appropriate product discussion, which helps us keep track of the demands for every feature: https://github.com/meilisearch/product/discussions/255

Also, over there, you can see that we had multiple ideas in mind; feel free to let us know what seems the best to you:
- `first`: If no document contains all the given terms, the engine removes the first query term, and so on.
- `optional`: All words in a query are always optional;
- `frequency`: If no document contains all the given terms, the engine removes the most common term within the dataset, and so on.
- `size`: If no document contains all the given terms, the engine removes the shorter term and so on.

-----

And finally, since there was some drama in this issue, just for context, I’m going to say that;
1. This week, we were on an offsite with the whole company and didn't really have the time to read this issue. That happens.
2. Quentin can answer whatever he wants; he doesn't need to answer your question first because you posted first
3. Then, as we are a company, we don't work on weekends, and at meilisearch, we work [four days per week, which means Fridays are included in our weekends](https://blog.meilisearch.com/how-we-implemented-the-4-day-work-week/).

**Comment by curquiza** [comment]
Hello everyone, sorry for the delay. As @irevoire explained, it was a special week for Meili team last week. We enjoyed time together and took the opportunity to work on multiple things we cannot do easily remotely.

It's not because we are full-time employed, we don't have priorities (we cannot be 100% dedicated to support), and we don't deserve respect. We cannot allow this kind of behavior in our repositories and in our community so we banned him, and we will ban any other person who does not follow our [code of conduct](https://github.com/meilisearch/meilisearch/blob/main/CODE_OF_CONDUCT.md)

Also, for those wondering: we tag all our issues as `support` first. Once we are sure it's a bug (and not related to any enhancements we should add to the engine) we finally tag it to `bug`. This makes sure we don't lose track of them, in our repository management. But don't worry, we don't consider an issue tagged as `support` less important than `bug` one!

### **Now, let's talk technically about the issue, technically**

As @CaroFG the issue is about typo-tolerance settings.
By default the typotolerance of Meilisearch works on words containing **at least 5 characters or more**. This default settings is here to avoid performance issue, but you can change it to match your use case!

See the default settings of Meilisearch:
```json
"typoTolerance": {
        "enabled": true,
        "minWordSizeForTypos": {
            "oneTypo": 5,
            "twoTypos": 9
        },
        "disableOnWords": [],
        "disableOnAttributes": []
    },
```

`appl` is 4 characters, so the typo-tolerance is not applied on it!

If you update your typo tolerance settings like this:

```json
{
    "minWordSizeForTypos": {
        "oneTypo": 4
    }
}
``` 

Searching for `appl macbook` will return the first document 👍 

Let me know if I miss something with the use case, I made my tests with the basic example that has been provided
---

@azngeek do you really have the same issue? Can you describe it so that we can help you?

**Comment by azngeek** [comment]
- I checked the case about the typo tolerance. The question is, how we can have typo tolerance about multiple search terms. So for example, if the search term is "Apple Macbook Pro" and the client-user for example types "appl macbok po", then I would expect Meilisearch to match at least for the terms "apple", "macbook" and "pro".

As the current matching strategy is "last" as default, what Elasticsearch does now is:

- Search for appl macbok po 
- Search for appl macbok
- Search for appl

And as the typo tolerance only applies to a single word and multiple words, separated with spaces are considered as an typo, i think that a solution would be to allow more than two typos? Just trying to understand the exact mechanics behind this.

`Concatenation When considering possible candidates for typo tolerance, Meilisearch will concatenate multiple search terms separated by a [space separator](https://www.meilisearch.com/docs/learn/advanced/datatypes#string). This is treated as one typo. For example, a search for any way would match documents containing anyway.`

**Comment by khalibloo** [comment]
Thanks for taking the time to explain how the matching works. It seems like it's not a bug then. It works as that strategy was intended. I will try tweaking the strategy and typo tolerance to see if I can get it to work to my expectation, before adding to the product discussion.

I'd also like to see what strategy fusejs.io uses. It is very good and very intuitive. I'm not very technical about the internals of search engines, but I'd very much like for the default strategy to be intuitive like that.

**Comment by xamir82** [comment]
Hey, I have the same problem. It would be nice if better matching strategies were added.
As other have noted, I, too, was expecting Meilisearch to operate on a token-by-token basis in these scenarios, so that if the entire query doesn't match any documents, it resorts to looking for individual words, while also applying typo-tolerance to each one. I think the `frequency` option that @irevoire describes is the most desirable approach.

@curquiza Your proposed solution of tweaking the typo tolerance settings might fix the issue in _this very specific example_, but there are other scenarios where the same problem sneaks in and those, @aradalvand alluded to and @irevoire confirmed, are about the matching strategy.

**Comment by azngeek** [comment]
- Maybe if we shift the discussion away from the technical capabilities to the user requirements, it will be clearer. As i am not 100 percent sure, how the things work, let us stay at the movie example:

Let's say, i want to find the movie "Testament of Youth". In this case the movie can be found when you type "testament" or also "testament ou youth".

But if you type "test youth", then you will not find it. But for each search string "test" and "youth", it will be displayed. I think that the expectation from a user is that there will be a matching also on this partial search strings. 

Does this explanation makes sense?

<img width="939" alt="image" src="https://github.com/meilisearch/meilisearch/assets/213429/94aae110-0931-4aba-8035-d9ae564ec175">

**Comment by curquiza** [comment]
Thank you very much @azngeek for your explanation, clearer indeed. I will report this to our PMs!

I let the issue open to ensure we got all the information from @xamir82 before closing the issue, which is so far more "feature request" oriented

**Comment by xamir82** [comment]
Thank you guys for you responses and your patience! I appreciate it.

Let me try to provide some concrete examples:

Imagine you have a document with a string field `Robert Smith`, if somebody searches for `Dr Smith`, no results are currently returned. I would expect Meilisearch to behave in such a way that if it doesn't find an exact match for the query (`Dr Smith`, in this case), it starts to look for individual words (`Smith`, in this example). I'm getting the expected results in Typesense in this case but Meilisearch doesn't return anything.
Correct me if I'm wrong but it seems like the `optional` matching strategy would actually solve this, right? But I'm just worried about `optional` resulting in too many hits, although if the hits are then sorted by "most words shared with the query" then that won't be a problem, will that be the case? I just wanted to ask to make sure, if so, then great.

Another example which is similar to the one @azngeek brought up, is that currently if you search for `shaw redemption` in the search bar on Melisearch's site home page ([meilisearch.com](https://meilisearch.com), it doesn't return Shawshank Redemption, but you would expect it to. This probably pertains to the "every token as prefix" thing that @irevoire mentioned, tracked by https://github.com/meilisearch/product/discussions/640. Is that correct?
But what about `shank redemption`? I think that won't be covered by the "every token as prefix" feature, this would require a "postfix" search, essentially. Is there an issue/discussion that tracks this one as well? I think that's important too.

@curquiza Thank you very much for your explanation! I didn't actually know typo-tolerance was already word-based, I guess the the lack of results in some of these examples which were due to the matching strategy and prefix search deceived me into thinking that it's because Meilisearch doesn't apply typo tolerance on a word-by-word basis. Good to know that that's not the issue, thanks for clarifying! Although I just tested and found out that if you search for `ap macbook pro` (even with your suggested typo-tolerance settings) it still doesn't return the documents, but I think this would be totally solved with the introduction of the `optional` matching strategy.

If these three issues are already tracked elsewhere then we can probably close this issue. Just one last thing though: Approximately when could we expect the `optional`/`frequency` and "every word as prefix" features to be added to Meilisearch? I totally understand that you guys probably have a lot on your plate and are very busy, but I think these are pretty essential features. I just want to know whether we're talking about weeks or months, so that I could decide whether I could use Meilisearch or not, I really love how clean the API and many other things about Meilisearch, but this is kind of a big issue for me. Could you perhaps give us a rough time frame? No pressure, just want to know if it's a matter of weeks, months, years? Thank you very much advance!

**Comment by curquiza** [comment]
Hello @xamir82
Thank you both with @azngeek and you for your feedback. This is really appreciated. There is a lot in this issue, this is really interesting!


> Imagine you have a document with a string field `Robert Smith`, if somebody searches for `Dr Smith`, no results are currently returned. I would expect Meilisearch to behave in such a way that if it doesn't find an exact match for the query (`Dr Smith`, in this case), it starts to look for individual words (`Smith`, in this example). I'm getting the expected results in TypeSense in this case but Meilisearch doesn't return anything.
> Correct me if I'm wrong but it seems like the `optional` matching strategy would actually solve this, right? But I'm just worried about `optional` resulting in too many hits, although if the hits are then sorted by "most words shared with the query" then that won't be a problem, will that be the case? I just want to ask to make sure, if so, then great.

Indeed, `optional` would fix this case. Or `frequency`, which could be a "smarter" `optional`. Nothing is defined and definitive yet. The purpose is to build the engine with users, so feel free to comment on [this discussion](https://github.com/meilisearch/product/discussions/255) to let us know about your detailed use case. I already linked this issue to the discussion, but your input will always be valuable in the discussion directly. 

More context about our organization if you are interested: the `product` repository is fully reserved for talking with users about their needs. This current repo (`meilisearch`) is more about our bugs and planned implementation even if we would never refuse to talk to users like here, of course, we will encourage them to go to the product repository.

> Another example which is similar to the one @azngeek brought up, is that currently if you search for `shaw redemption` in the search bar on the Melisearch home page ([meilisearch.com](https://meilisearch.com)), it doesn't return Shawshank Redemption, but you would expect it to. This probably pertains to the "every token as prefix" thing that @irevoire mentioned, tracked by [meilisearch/product#640](https://github.com/meilisearch/product/discussions/640). Is that correct?

In this case, this is still related to relevancy but not internally the same for Meilisearch. Indeed, you are correct, this is more related to https://github.com/meilisearch/product/discussions/640 and https://github.com/meilisearch/product/discussions/97#discussioncomment-2014212, so apply the prefix on each query word. I'm not sure, but it would depend on our final implementation, the `optional` or `frequency` option would help for this.
Again, feel free to comment on the discussion. 

> But what about `shank redemption`? I think that won't be covered by the "every token as prefix" feature, this would require a "postfix" search, essentially. Is there an issue/discussion that tracks this one as well? I think that's important too. Although you might argue that adding `optional` as a matching strategy could solve some of these cases as well.

This one is another one again! 😄 It's about finding part of a word, so disabling prefix search. You are not the first one asking for this. Here is the related product discussion: https://github.com/meilisearch/product/discussions/492
This one is a little bit trickier for Meilisearch since it goes completely against the design of the engine, that has been created to perform prefix search. And this one could not be compatible with "enable prefix search on each quary term".

> @curquiza Thank you very much for your explanation! I didn't actually know typo-tolerance was already word-based, I guess the the lack of results in some of these examples which were due to the matching strategy and prefix search deceived me into thinking that it's because Meilisearch doesn't apply typo tolerance on a word-by-word basis. Good to know that that's not the issue, thanks for clarifying! Although I just tested and found out that if you search for `ap macbook pro 2019` (even with your suggested typo-tolerance settings) it still doesn't return the documents even though most of the words are there and they're typo-free, but I think this once again the sort of thing that the introduction of the `optional` matching strategy would solve.

No problem! Meilisearch is a complex software, it's normal to not have all the information when asking questions and concerns.
For the `ap macbook pro 2019`, it indeed does not work. And it would work with `optional`/`frequency`.


> If these three issues are already tracked elsewhere then we can close this issue. Just one last thing though: Approximately when could we expect the `optional`/`frequency` and "every word as prefix" features to be added to Meilisearch? I totally understand that you guys probably have a lot on your plate and are very busy, but I think these are pretty essential features. I just want to know whether we're talking about weeks or months, so that I could decide whether or not I could "bet on" Meilisearch now, so to speak, and wait for these issues to soon be fixed. Cause I really love how clean the API is, the documentation, it being Rust-based, and many other things about Meilisearch, but this is kind of a big issue for me. Would it be possible for you to give us a rough time frame? No pressure, just want to know if this is weeks away, months away, years away? Hopefully it's prioritized! Thank you very much in advance!

Since these issues are still in the discovery stage, the issues are tracked in the different `product` repository discussions I linked. 
To be fully transparent with you, the team cannot really focus on this now. We have indeed already set our priority for the quarter and we have a lot on our plate: HA (High Avaibility), reducing indexing speed, promoting & boosting and improving our vector search. But this can definitely be something we can consider for the beginning of 2024, especially `optional`/`frequency`! 😊
In the meantime, for the `optional`, you have a workaround by doing a [multi-search on each query term](https://www.meilisearch.com/docs/reference/api/multi_search#perform-a-multi-search) as @azngeek described, but I know it's far away from being convenient.

---


TLDR; for our PMs @gmourier and @macraig 

- Looks like the biggest issue: users are missing relevancy options, like `optional`/`frequency`. Related [product discussion](https://github.com/meilisearch/product/discussions/255)
- Users want to activate prefix search on each query term. Related [product discussion](https://github.com/meilisearch/product/discussions/640).
- Users also want to fully disable prefix search and want the query term to match the sub-part of a word, but not the beginning. Like finding `github` when typing `hub`. Related [product discussion](https://github.com/meilisearch/product/discussions/492).


Again thank you everyone for your help. I think I can close the issue now! Feel free to come to us on the `product` repo (or in this one if you find bugs, but I hope you will not 😇)
Have a nice day 👋

**Comment by bedus-creation** [comment]
> ll the search term combination in multiple search requests using multi-search and

Did you implement pagination as well with multi search ?

**Comment by AnirudhaGohokar** [comment]
The  matching strategy specially `last` is generating incorrect results and really want to know why this was implemented.

For example just swithching the search terms would generate incorrect or missing results using last strategy:
```
aiml ml
ml aiml
```

I think the matching strategy should be replaced with fieldValueOperator whose values could be AND , OR.
The AND operator will match all keywords in the given record which is similar to your "ALL" operator.
The OR operator will optionally match all keywords in given record.

Similarly this can implemented across the fields i.e fieldOperator for multiple fields and fieldValueOperator for fields value.

We were planning to replace Elastic Search with Milisearch but it seems right now its not even proper state. Maybe in future once it is stable enough surely it could outperform Elastic Search.

Keep up the good work!!

**Comment by ManyTheFish** [comment]
Hello @AnirudhaGohokar,
As you said, Meilisearch doesn't provide this kind of feature.
However, did you consider using [the `frequency` matching strategy](https://www.meilisearch.com/docs/reference/api/search#frequency)? This strategy may fit your needs.
Let me know after trying it if it does,
thanks!
